### ISSUE TITLE
WARNING:root:SoundDeviceStream write underflow (size: 1600)

### ISSUE THREAD

**Issue by Pillar1989** [googlesamples/assistant-sdk-python#12]
Hi
I verify audio setup works fine.
But when I run the demo, there are too many "WARNING:root:SoundDeviceStream write underflow (size: 1600)",  record and play are discontinuity.
```
(env)pi@raspberrypi:~ $ proxychains python -m googlesamples.assistant   
ProxyChains-3.1 (http://proxychains.sf.net)
|S-chain|-<>-127.0.0.1:1080-<><>-4.2.2.2:53-<><>-OK
|S-chain|-<>-127.0.0.1:1080-<><>-4.2.2.2:53-<><>-OK
INFO:root:Connecting to embeddedassistant.googleapis.com
Press Enter to send a new request...
INFO:root:Recording audio request.
|S-chain|-<>-127.0.0.1:1080-<><>-4.2.2.2:53-<><>-OK
WARNING:root:SoundDeviceStream read overflow (3200, 6400)
INFO:root:End of audio request detected
INFO:root:Transcript of user request: "what's the timer block".
INFO:root:Playing assistant response.
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
INFO:root:Finished playing assistant response.
Press Enter to send a new request...
INFO:root:Recording audio request.
ERROR:root:server error: Service unavailable.
INFO:root:Finished playing assistant response.
Press Enter to send a new request...
INFO:root:Recording audio request.
INFO:root:End of audio request detected
INFO:root:Transcript of user request: "how are you".
INFO:root:Playing assistant response.
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
INFO:root:Finished playing assistant response.
Press Enter to send a new request...
INFO:root:Recording audio request.
^C
Aborted!
```

**Comment by proppy** [comment]
Which microphone / speaker are you using?

Do you have the same problem when running `python -m googlesamples.assistant.audio_helpers` ?

**Comment by Pillar1989** [comment]
python -m googlesamples.assistant.audio_helpers works fine. I didnt have any problem.

**Comment by proppy** [comment]
And the assistant sample doesn't?

Can you clarify the hardware you're using:
- Architecture and OS?
- Microphone type?
- Speaker type?

**Comment by Pillar1989** [comment]
We use seeed voice card. 
```
pi@raspberrypi:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Raspbian
Description:    Raspbian GNU/Linux 8.0 (jessie)
Release:        8.0
Codename:       jessie
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.9.13-v7+ #974 SMP Wed Mar 1 20:09:48 GMT 2017 armv7l GNU/Linux
pi@raspberrypi:~ $ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: seeedvoicecard [seeed-voicecard], device 0: bcm2835-i2s-wm8960-hifi wm8960-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
pi@raspberrypi:~ $ arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: seeedvoicecard [seeed-voicecard], device 0: bcm2835-i2s-wm8960-hifi wm8960-hifi-0 []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
pi@raspberrypi:~ $ 
```

**Comment by proppy** [comment]
> We use seeed voice card.

Can you link to the product page? Thanks.

**Comment by Pillar1989** [comment]
Sorry, seeed-voicecard have not yet been released. 
But I verify python -m googlesamples.assistant.audio_helpers works fine.

**Comment by proppy** [comment]
Can you try to run the sample with different `--audio-block-size` values ?

Try:
```
python -m googlesamples.assistant  --audio-block-size=0
python -m googlesamples.assistant  --audio-block-size=1024
python -m googlesamples.assistant  --audio-block-size=2048
python -m googlesamples.assistant  --audio-block-size=4096
```

**Comment by Pillar1989** [comment]
I dont know what's wrong,  when I run   --audio-block-size option. I didnt see "WARNING:root:SoundDeviceStream write underflow (size: 1600)" . and when I remove --audio-block-size option, retest It. The WARNING  disappeared.

**Comment by proppy** [comment]
@Pillar1989 can you also share the content of your `.asoundrc` ?

**Comment by Pillar1989** [comment]
```
(env)pi@raspberrypi:~ $ cat .asoundrc 
pcm.!default {
  type asym
  capture.pcm "mic"
  playback.pcm "speaker"
}
pcm.mic {
  type plug
  slave {
    pcm "hw:1,0"
  }
}
pcm.speaker {
  type plug
  slave {
    pcm "hw:1,0"
  }
}
```

**Comment by mthoretton** [comment]
Hey,
I tried to play with `--audio-iter-size` and `--audio-block-size` but it seems that doesn't have any influence on the issue. I'm using a raspi 3 with a "JBL Go" as mic/speaker through bluetooth (using an external usb/bluetooth adapter).

**Comment by proppy** [comment]
@mthoretton can you open a separate issue with point to your microphone and speaker?

**Comment by Pillar1989** [comment]
Thanks @proppy

**Comment by mikeleonard** [comment]
@Pillar1989 @proppy 

Hi. Just FYI, I also have the same ```WARNING:root:SoundDeviceStream write underflow (size: 6144)``` issue when running the python code on my macbook pro. 

I tried doubling and halving all the numeric settings e.g. audio-iter-size/audio-block-size/etc and these didn't seem to work. 

Note that: it still seems to work fine (I mean I can converse with google assistant).

I think I've worked out what's causing the problem: when I run with --verbose, it's clear that the underflow message is printed only after an END_OF_UTTERANCE message. Specifically, I get it when the code sends sends off an ```audio_in``` message AFTER receiving the end of utterance message. 

```
...
DEBUG:root:ConverseRequest: audio_in (6400 bytes)
DEBUG:root:ConverseRequest: audio_in (6400 bytes)
DEBUG:root:ConverseRequest: audio_in (6400 bytes)
DEBUG:root:ConverseRequest: audio_in (6400 bytes)
DEBUG:root:ConverseRequest: audio_in (6400 bytes)
DEBUG:root:ConverseResponse: event_type: END_OF_UTTERANCE

INFO:root:End of audio request detected
DEBUG:root:ConverseResponse: result {
  spoken_request_text: "what\'s the time"
}

INFO:root:Transcript of user request: "what's the time".
INFO:root:Playing assistant response.
DEBUG:root:ConverseRequest: audio_in (6400 bytes)   # sent AFTER receiving END_OF_UTTERANCE
DEBUG:root:ConverseResponse: result {
  conversation_state: "\n&C#5911....xMTRlZjViNBgB"
  microphone_mode: CLOSE_MICROPHONE
}

DEBUG:root:ConverseResponse: audio_data (1600 bytes)
WARNING:root:SoundDeviceStream write underflow (size: 1600) # note the size here isn't always 1600, I've seen larger values.
DEBUG:root:ConverseResponse: audio_data (1600 bytes)
DEBUG:root:ConverseResponse: audio_data (1600 bytes)
DEBUG:root:ConverseResponse: audio_data (1600 bytes)
DEBUG:root:ConverseResponse: audio_data (1600 bytes)
...
```

**Comment by Klhnikov** [comment]
Same problem here on PC with grpc.pushtotalk : 

> INFO:root:Playing assistant response.
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
WARNING:root:SoundDeviceStream write underflow (size: 1600)
INFO:root:Finished playing assistant response.
Press Enter to send a new request...`
